### PR TITLE
Report generating feature

### PR DESCRIPTION
**Pull request by kts-desilva**
## Purpose
This PR introduces functionality to export the dashboard and its widgets as pdf files. The charts and the overall dashboard view are exported as images and the tables are exported in formated table format.

## Goals
To export the dashboard as pdf.
Export the widgets separately as pdf.

## Approach
The dashboard can be donwloaded as a pdf. Preferred pages in the dashboard can be exported by 'capture current page' button in the side pane. Then the page size can be selected and export the pages as a single pdf file. If the user wishes to include the report generation time, it also can be included in the report. It is prompted once after the export button is pressed. If the user wants to remove a page from the report it can be done.

The widgets can be exported seperately as pdf files. There also it is promoted to include the report generation time. The table chart type is exported in the table format and there user can select the number of records to be included in the report. Rest of the chart types are exported as snapshot images.

Please refer [1] for more details.

[1] [https://docs.google.com/document/d/13E73LuVD5kE511ue2Vfx1MhSPj4Ve15_JQh6SV2by_Q/edit?](url)

### CODE REVIEW

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 1]
```diff
@@ -272,6 +418,234 @@ class DashboardViewPage extends Component {
...
+        const docName = title + ".pdf";
+        pdf.save(docName);
+
+    }
+
+    // Get the image from deployment.yaml file through the server
+    async addPdfImage(url, callback) {
+        let path = "/portal/public/app/images/"
+
+        await axios.get(url, { auth: { username: "admin", password: "admin" } })
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Any reason to hard code passwords?

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 2]
```diff
@@ -272,6 +418,234 @@ class DashboardViewPage extends Component {
...
+                    const canvas = document.createElement('canvas');
+                    const ctx = canvas.getContext('2d');
+                    const img = new Image();
+                    let imgStr;
+
+                    img.onload = function () {
+                        canvas.width = img.width;
+                        canvas.height = img.height;
+                        ctx.drawImage(img, 0, 0);
+                        imgStr = canvas.toDataURL("image/png");
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Use single quotes and double quotes as per the conventions. Please change other places as well.

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 3]
```diff
@@ -272,6 +418,234 @@ class DashboardViewPage extends Component {
...
+
+                    img.src = path;
+                }
+
+            });
+    }
+
+    async removePage(index) {
+
+        const newAr = this.state.pageList.slice();
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Use meaningful names

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 4]
```diff
@@ -272,6 +418,234 @@ class DashboardViewPage extends Component {
...
+            newAr.push(currentPage);
+            this.setState({ pageList: newAr });
+
+            const val = 620 / canvas.width;
+            this.state.resizeRatio = val;
+            this.state.canvasWidth = canvas.width;
+            this.state.canvasHeight = canvas.height;
+        });
+    }
+
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: DashboardViewPage.jsx is too long. Let's try to break into Utill classes

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx [hunk 5]
```diff
@@ -182,6 +271,322 @@ export default class DashboardRenderer extends Component {
...
+            const totalPagesExp = "{total_pages_count_string}";
+
+            await this.addPdfImage("/portal/apis/dashboards/pdfHeader", function (res) {
+
+                localStorage.setItem("dashboardHeader", res);
+            });
+
+            await this.addPdfImage("/portal/apis/dashboards/pdfFooter", function (res) {
+                if (res != 'none') {
+
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: remove unnecessary new lines

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx [hunk 6]
```diff
@@ -182,6 +271,322 @@ export default class DashboardRenderer extends Component {
...
+                    row.push(rowDivs[colNum * i + j].innerText)
+                }
+                rowData.push(row);
+            }
+
+            const doc = new jspdf('p', 'pt');
+            const totalPagesExp = "{total_pages_count_string}";
+
+            await this.addPdfImage("/portal/apis/dashboards/pdfHeader", function (res) {
+
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: remove unnecessary new lines

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx [hunk 7]
```diff
@@ -182,6 +271,322 @@ export default class DashboardRenderer extends Component {
...
+                // HEADER
+                const headerImg = localStorage.getItem('dashboardHeader');
+                if (headerImg != null) {
+                    doc.addImage(headerImg, "JPEG", data.settings.margin.left,
+                                 pdfConfig.stampImageDashboard.coordinates.y, pdfConfig.stampImageDashboard.size.x,
+                                 pdfConfig.stampImageDashboard.size.y);
+                }
+
+                // FOOTER
+                let str = "Page " + data.pageCount;
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: use meaningful names

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx [hunk 8]
```diff
@@ -182,6 +271,322 @@ export default class DashboardRenderer extends Component {
...
+            this.state.includeRecords = false;
+            this.state.includeTime = false;
+
+            const tableDocumentName = docTitle + ".pdf";
+            doc.save(tableDocumentName);
+
+
+        } else {
+            //Add the element's snapshot into the page
+            await this.addPdfImage("/portal/apis/dashboards/pdfHeader", function (res) {
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: avoid hard coding app context

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx [hunk 9]
```diff
@@ -182,6 +271,322 @@ export default class DashboardRenderer extends Component {
...
+            this.state.includeRecords = false;
+            this.state.includeTime = false;
+
+            const documentName = docTitle + ".pdf";
+            pdf.save(documentName);
+        }
+    }
+
+    async addPdfImage(url, callback) {
+        let path = "/portal/public/app/images/"
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: avoid hard coding appcontext. Please change else where.

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx [hunk 10]
```diff
@@ -182,6 +271,322 @@ export default class DashboardRenderer extends Component {
...
+        this.setState({ reportStatusOpen: true });
+    }
+
+    progress = () => {
+        let { completed } = this.state;
+        if(isNaN(completed)){
+            completed=0;
+        }
+        this.setState({ completed: completed >= 100 ? 0 : completed + 10 });
+    };
```
Comment (TanyaM) [inline components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: If there are any public facing strings, localize them

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 11]
```diff
@@ -272,6 +418,234 @@ class DashboardViewPage extends Component {
...
+        this.handleClose();
+        this.handleReportStatusOpen();
+        this.timer = setInterval(this.progress, 195);
+
+        let pdf = new jspdf("l", "px", this.state.pageSize.toLowerCase());
+
+        const title = document.getElementsByTagName("h1")[0].innerText;
+
+        // Get header : a cutomized header defined by the user or else the default stream processor header
+        await this.addPdfImage("/portal/apis/dashboards/pdfHeader", function (res) {
```
[the remark this answers is not in the document]
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Done

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 12]
```diff
@@ -272,6 +418,234 @@ class DashboardViewPage extends Component {
...
+            if (res != 'none') {
+                localStorage.setItem("dashboardHeader", res);
+            }
+        });
+
+        // Get footer : a cutomized footer defined by the user or else the no footer by default
+        await this.addPdfImage("/portal/apis/dashboards/pdfFooter", function (res) {
```
[the remark this answers is not in the document]
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Done

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 13]
```diff
@@ -272,6 +418,234 @@ class DashboardViewPage extends Component {
...
+        this.state.includeTime = false;
+
+        const docName = title + ".pdf";
+        pdf.save(docName);
+
+    }
+
+    // Get the image from deployment.yaml file through the server
+    async addPdfImage(url, callback) {
+        let path = "/portal/public/app/images/"
```
[the remark this answers is not in the document]
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Done

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx (hunk 6 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: Done

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx (hunk 7 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: Done

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx (hunk 8 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: Done

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx (hunk 9 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: Done

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx (hunk 2 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Done

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx (hunk 3 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Done

File: components/dashboards-web-component/src/viewer/components/dashboard-container-styles.css [hunk 14]
```diff
@@ -24,4 +24,13 @@
     margin: 0;
     border: 0;
 }
-
+.widget-export-button {
+    float: left;
+    color: #929292;
```
Comment (this) [inline components/dashboards-web-component/src/viewer/components/dashboard-container-styles.css]: Don't declare custom colors, instead use theme colors.

File: components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx (hunk 10 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/components/DashboardRenderer.jsx]: Done

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx (hunk 4 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Done

**Comment by this** [comment]
@kts-desilva
Since we need to add this to the SP-4.3.0-alpha release, first fix the merge conflict. Then we can merge this PR. After that you can work on above suggested refactoring.

File: components/dashboards-web-component/src/viewer/components/dashboard-container-styles.css (hunk 14 above)
Comment (kts-desilva) [reply components/dashboards-web-component/src/viewer/components/dashboard-container-styles.css]: Done

File: components/dashboards/org.wso2.carbon.dashboards.core/src/main/java/org/wso2/carbon/dashboards/core/internal/DashboardMetadataProviderImpl.java [hunk 15]
```diff
@@ -527,4 +527,25 @@ protected void unsetIdP(IdPClient client) {
         this.identityClient = null;
     }
 
+    public String getPdfHeader() {
+        try {
+            Object x = ((Map<String, Object>) configProvider.getConfigurationObject("wso2.status.dashboard"))
+                    .get("pdfHeader");
+            return x.toString();
+        } catch (ConfigurationException e) {
+            LOGGER.error("Error in getting the pdf header image", e);
```
Comment (this) [inline components/dashboards/org.wso2.carbon.dashboards.core/src/main/java/org/wso2/carbon/dashboards/core/internal/DashboardMetadataProviderImpl.java]: Why consume configurations of `wso2.status.dashboard`? 

- We shouldn't consume configurations from other namespaces, only from `wso2.dashboard` namespace.
- Don't declare dashboard configuration arbitrarily. Dashboard configurations from `deployment.yaml` is loaded into [DashboardConfigurations](https://github.com/wso2/carbon-dashboards/blob/master/components/dashboards/org.wso2.carbon.dashboards.core/src/main/java/org/wso2/carbon/dashboards/core/bean/DashboardConfigurations.java) bean class. If you are introducing a new configuration, then that should be added into the `DashboardConfigurations` class.

Please fix this in the next PR.

File: components/dashboards/org.wso2.carbon.dashboards.core/src/main/java/org/wso2/carbon/dashboards/core/internal/DashboardMetadataProviderImpl.java [hunk 16]
```diff
@@ -527,4 +527,25 @@ protected void unsetIdP(IdPClient client) {
...
+        }
+        return "";
+    }
+
+    public String getPdfFooter() {
+        try {
+            Object x = ((Map<String, Object>) configProvider.getConfigurationObject("wso2.status.dashboard"))
```
Comment (this) [inline components/dashboards/org.wso2.carbon.dashboards.core/src/main/java/org/wso2/carbon/dashboards/core/internal/DashboardMetadataProviderImpl.java]: Above comment applies here as well.

**Pull request merged**